va/caa.go

```diff
@@ -63,85 +63,65 @@ func (va *ValidationAuthorityImpl) checkCAA(
 	}
 
-	va.log.AuditInfof("Checked CAA records for %s, [Present: %t, Account ID: %d, Challenge: %s, Valid for issuance: %t] Response=%q",
-		identifier.Value, present, params.accountURIID, params.validationMethod, valid, response)
+	va.log.AuditInfof("Checked CAA records for %s, [Present: %t, Account ID: %d, Challenge: %s, Valid for issuance: %t, Found at: %q] Response=%q",
+		identifier.Value, foundAt != "", params.accountURIID, params.validationMethod, valid, foundAt, response)
 	if !valid {
-		return probs.CAA(fmt.Sprintf("CAA record for %s prevents issuance", identifier.Value))
+		return probs.CAA(fmt.Sprintf("CAA record for %s prevents issuance", foundAt))
 	}
 	return nil
 }
 
-// CAASet consists of filtered CAA records
-type CAASet struct {
-	Issue     []*dns.CAA
-	Issuewild []*dns.CAA
-	Iodef     []*dns.CAA
-	Unknown   []*dns.CAA
+// caaResult represents the result of querying CAA for a single name. It breaks
+// the CAA resource records down by category, keeping only the issue and
+// issuewild records. It also records whether any unrecognized RRs were marked
+// critical, and stores the raw response text for logging and debugging.
+type caaResult struct {
+	name            string
+	issue           []*dns.CAA
+	issuewild       []*dns.CAA
+	criticalUnknown bool
+	raw             string
+	err             error
 }
 
-// returns true if any CAA records have unknown tag properties and are flagged critical.
-func (caaSet CAASet) criticalUnknown() bool {
-	if len(caaSet.Unknown) > 0 {
-		for _, caaRecord := range caaSet.Unknown {
+// filterCAA processes a set of CAA resource records and picks out the only bits
+// we care about: the issue records, the issuewild records, and whether any
+// unrecognized records had the critical bit set.
+func filterCAA(rrs []*dns.CAA) ([]*dns.CAA, []*dns.CAA, bool) {
+	var issue, issuewild []*dns.CAA
+	var criticalUnknown bool
+
+	for _, caaRecord := range rrs {
+		switch strings.ToLower(caaRecord.Tag) {
+		case "issue":
+			issue = append(issue, caaRecord)
+		case "issuewild":
+			issuewild = append(issuewild, caaRecord)
+		default:
 			// The critical flag is the bit with significance 128. However, many CAA
 			// record users have misinterpreted the RFC and concluded that the bit
 			// with significance 1 is the critical bit. This is sufficiently
 			// widespread that that bit must reasonably be considered an alias for
 			// the critical bit. The remaining bits are 0/ignore as proscribed by the
 			// RFC.
 			if (caaRecord.Flag & (128 | 1)) != 0 {
-				return true
+				criticalUnknown = true
 			}
 		}
 	}
 
-	return false
-}
-
-// Filter CAA records by property
-func newCAASet(CAAs []*dns.CAA) *CAASet {
-	var filtered CAASet
-
-	for _, caaRecord := range CAAs {
-		switch strings.ToLower(caaRecord.Tag) {
-		case "issue":
-			filtered.Issue = append(filtered.Issue, caaRecord)
-		case "issuewild":
-			filtered.Issuewild = append(filtered.Issuewild, caaRecord)
-		case "iodef":
```

Comment:
Yep! It would be pretty easy to add back iodef records to this struct in the future, but for now we don't do anything with them, so imo there's no reason to have dead code carrying them around.